src/core/sorted_map.h

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+
+#include <absl/functional/function_ref.h>
+
+#include <optional>
+#include <string>
+#include <vector>
+
+extern "C" {
+#include "redis/dict.h"
+#include "redis/zset.h"
+}
+
+namespace dfly {
+namespace detail {
+
+/**
+ * @brief SortedMap is a sorted map implementation based on zset.h. It holds unique strings that
+ * are ordered by score and lexicographically. The score is a double value and has higher priority.
+ * The map is implemented as a skip list and a hash table. For more details see
+ * zset.h and t_zset.c files in Redis.
+ */
+class SortedMap {
+ public:
+  using ScoredMember = std::pair<std::string, double>;
+  using ScoredArray = std::vector<ScoredMember>;
+
+  SortedMap();
+  SortedMap(const SortedMap&) = delete;
+  SortedMap& operator=(const SortedMap&) = delete;
+
+  SortedMap(SortedMap&& other) noexcept = default;
+  SortedMap& operator=(SortedMap&& other) noexcept = default;
```

Comment:
Actually I don't think that the default move operations are a good fit here, as they won't `null`ify the members of the moved-from object.
If this class, like you implied, is only ever used on the heap then you can `delete` those as well. If you do think it might be useful to have these movable, then I think you'll have to implement these operations yourself, and make sure that in case the members are `nullptr` the d'tor works as expected.

Reply:
good catch